dbms/src/Flash/Mpp/ExchangeReceiver.cpp

```diff
@@ -914,48 +988,49 @@ void ExchangeReceiverBase<RPCContext>::connectionDone(
     const String & local_err_msg,
     const LoggerPtr & log)
 {
-    Int32 copy_live_connections;
     String first_err_msg = local_err_msg;
-    {
-        std::lock_guard lock(mu);
 
-        if (meet_error)
-        {
-            if (state == ExchangeReceiverState::NORMAL)
-                state = ExchangeReceiverState::ERROR;
-            if (err_msg.empty())
-                err_msg = local_err_msg;
-            else
-                first_err_msg = err_msg;
-        }
+    // We must protect the following codes with lock.
+    // Because once live_connections == 0, ExchangeReceiver may be destructed immediately
+    // and the data in ExchangeReceiver is no longer alive.
```

Comment:
Or like `local_tunnel_v2`, add a `live_async_connections`?

Reply:
or call `async_handler_ptrs.wait()` in `~Exchange()` just like `thread_manager.wait()`?

Reply:
> Or like `local_tunnel_v2`, add a `live_async_connections`?

okk, I will change it.